arangodb filer store

gfxlabs opened the pull request:
hey!

i was curious if i was on the right path for a filer store implementation for [arangodb](https://github.com/arangodb/arangodb/). we use this database a lot internally.

i tested it out and it seemed to be working, but that's just from me playing around and uploading some files

i plan to add bucketaware later. am I correct to believe that OnBucketDeletion should wipe everything with said bucket, and CanDropWholeBucket should return true?

chrislusf commented:
Thanks! This is great!

"i plan to add bucketaware later. am I correct to believe that OnBucketDeletion should wipe everything with said bucket, and CanDropWholeBucket should return true?" Yes

btw: The `ListDirectoryPrefixedEntries` is just listing entries under one directory with a prefix, similar to 
```
ls -l /path/to/dir/prefix*
```

gfxlabs commented on weed/filer/arangodb/arangodb_store.go:
```diff
@@ -0,0 +1,393 @@
+	if err != nil {
+		return err
+	}
+	db_name := "seaweed-filer"
```
for sure, i'll make some changes and finish this up

gfxlabs commented:
ok, i moved the db name to config and moved a few things around. 

i added the methods for the buckets, but i'm not sure if they are being called. do i need to run a cluster to test those? not 100% on the internals there

anyways, i have a filer with that ~3.5mm files, i'll try syncing that over and seeing if any issues pop up

chrislusf commented on weed/filer/arangodb/readme.md:
```diff
@@ -31,12 +31,12 @@ i test with
 ## thoughts
 
-should there be one collection per bucket? would make deleting a bucket instant as compared to fast
+should there be one collection per bucket? this would make deleting a bucket O(1) instead of O(n)
```
yes, this was the assumption.

gfxlabs replied:
gotcha, will do tn

chrislusf commented:
what is the status for this PR?

gfxlabs commented:
hey, 

sorry, i've been a bit busy the past few days. 

it's been working so far on my local computer with some basic operations + replication. i was going to find some times to test it on k8s, probably by the end of the week.

i'll update here again after i make sure no issues come up when testing on the cluster.

gfxlabs commented:
finally found some time to migrate the filer meta, sorry for the delay

i refactored your [seaweedfs operator](https://gfx.cafe/open/seaweedfs-operator) with some updates for it to work with k8s 1.22+ and 1.7.2 cert manager, 

it's currently running with this config, serving ~5mm files. i'm gonna just let this run and hope it doesn't break lol.

are there any integration tests i should run ?
```
apiVersion: seaweed.seaweedfs.com/v1
kind: Seaweed
metadata:
  name: seaweed
spec:
  enablePVReclaim: true
  filer:
    affinity:
      podAntiAffinity:
        preferredDuringSchedulingIgnoredDuringExecution:
        - podAffinityTerm:
            labelSelector:
              matchExpressions:
              - key: app.kubernetes.io/component
                operator: In
                values:
                - filer
            topologyKey: kubernetes.io/hostname
          weight: 100
    config: |
      [arangodb]
      enabled = true
      db_name = "seaweedfs"
      servers=["xxx"]
      username="seaweedfs"
      insecure_skip_verify =true
    env:
    - name: WEED_ARANGODB_PASSWORD
      valueFrom:
        secretKeyRef:
          key: xxx
          name: xxx
    replicas: 3
  hostSuffix: ""
  image: gfxlabs/seaweedfs:arango
  master:
    affinity:
      podAntiAffinity:
        preferredDuringSchedulingIgnoredDuringExecution:
        - podAffinityTerm:
            topologyKey: kubernetes.io/hostname
          weight: 100
    defaultReplication: "001"
    replicas: 3
    volumeSizeLimitMB: 16024
  metricsAddress: xxx
  volume:
    affinity:
      podAntiAffinity:
        preferredDuringSchedulingIgnoredDuringExecution:
        - podAffinityTerm:
            labelSelector:
              matchExpressions:
              - key: app.kubernetes.io/component
                operator: In
                values:
                - volume
            topologyKey: kubernetes.io/hostname
          weight: 100
    replicas: 10
    requests:
      storage: xxx
    storageClassName: xxx
  volumeServerDiskCount: 1

```

gfxlabs commented:
heya

the docker image build with this fork has been running on our prod for two weeks without any issues so far, the database seems to be performing well so i'm feeling confident

@chrislusf is there anything else i need to do/add for this pr?

chrislusf commented:
Thanks a lot!

Maybe help to add a section about it here: https://github.com/chrislusf/seaweedfs/wiki/Filer-Stores

Also, what are the benefits of [arangodb](https://github.com/arangodb/arangodb/) besides it is widely used in your company?

gfxlabs commented:
i can add a section there no issue.

arangodb is a generic document store, so just imagine mongodb but actually usable haha.... 
quite easy to setup and support is pretty good. the underlying storage is just rocksdb. we've replaced internal use of redis with arangodb as well.
it also supports sql-like joins between collections, letting us use the same single database for our row-based data also.
